src/Keboola/GoogleSheetsWriter/Sheet.php

```diff
@@ -38,36 +39,31 @@ public function process(array $sheet): array
             $sheetProperties = $this->getSheetProperties($sheet['fileId'], $sheet['sheetId']);
             $columnCount = $this->inputTable->getColumnCount();
             $rowCountSrc = $this->inputTable->getRowCount();
-
-            if (empty($sheetProperties)) {
-                throw new UserException(sprintf(
-                    'Sheet "%s" (%s) not found in file "%s" (%s)',
-                    $sheet['sheetTitle'],
-                    $sheet['sheetId'],
-                    $sheet['title'],
-                    $sheet['fileId']
-                ));
-            }
-
-            if ($columnCount * $rowCountSrc > 2000000) {
-                throw new UserException('CSV file exceeds the limit of 2000000 cells');
-            }
+            $this->preFlightChecks($sheet, $sheetProperties, $columnCount, $rowCountSrc);
 
             // update columns
             $rowCountDst = $sheetProperties['properties']['gridProperties']['rowCount'];
             $this->updateMetadata($sheet, ['columnCount' => $columnCount, 'rowCount' => $rowCountDst]);
 
-            if ($sheet['action'] === ConfigDefinition::ACTION_UPDATE) {
-                $this->client->clearSpreadsheetValues(
-                    $sheet['fileId'],
-                    $this->getRange($sheet['sheetTitle'], $columnCount, 1, $rowCountDst)
-                );
-                // update rows to match source size
-                $this->updateMetadata($sheet, ['columnCount' => $columnCount, 'rowCount' => $rowCountSrc]);
-            }
-
             // upload data
-            return $this->uploadValues($sheet, $this->inputTable);
+            switch ($sheet['action']) {
+                case ConfigDefinition::ACTION_UPDATE:
+                    $this->client->clearSpreadsheetValues(
+                        $sheet['fileId'],
+                        $this->getRange($sheet['sheetTitle'], $columnCount, 1, $rowCountDst)
+                    );
+
+                    // update rows to match source size
+                    $this->updateMetadata($sheet, ['columnCount' => $columnCount, 'rowCount' => $rowCountSrc]);
+
+                    return $this->updateAction($sheet, $this->inputTable);
+                case ConfigDefinition::ACTION_APPEND:
+                    return $this->appendAction($sheet, $this->inputTable);
+                    break;
+                default:
+                    throw new ApplicationException(sprintf("Action '%s' not allowed", $sheet['action']));
```

Comment:
Možná lepší "Unknown action '%s'". Takhle to navádí na to zkoumat jestli nejde nějak "povolit", když "není povolená".